### PR TITLE
ARROW-1578: [C++] Run lint checks in Travis CI much earlier at before_script stage to fail faster

### PR DESCRIPTION
**Pull request by wesm**

### CODE REVIEW

File: cpp/src/arrow/buffer.h [hunk 1]
```diff
@@ -173,7 +173,7 @@ class ARROW_EXPORT ResizableBuffer : public MutableBuffer {
 /// A Buffer whose lifetime is tied to a particular MemoryPool
 class ARROW_EXPORT PoolBuffer : public ResizableBuffer {
  public:
-  explicit PoolBuffer(MemoryPool* pool = nullptr);
+  PoolBuffer(MemoryPool* pool = nullptr);
```
Comment (xhochy) [inline cpp/src/arrow/buffer.h]: Why did you remove this?

Comment (wesm) [reply cpp/src/arrow/buffer.h]: This will fail the cpplint checks so I wanted to make sure that my changes have the desired effect of failing the builds fast

**Review by xhochy — APPROVED** [review]

**Comment by wesm** [comment]
Pretty unsure why https://travis-ci.org/apache/arrow/jobs/277763849 failed
